File: LeetCode (en comments)/912. Sort an Array.java
class Solution {
    public int[] sortArray(int[] nums) {
        if (nums.length < 2) {
            return nums;
        } else if (nums.length == 2) {
            if (nums[0] > nums[1]) {
                return new int[]{nums[1], nums[0]};
            } else {
                return nums;
            }
        }
        int[] left = Arrays.copyOfRange(nums, 0, nums.length / 2);
        int[] right = Arrays.copyOfRange(nums, nums.length / 2, nums.length);
        left = sortArray(left);
        right = sortArray(right);
        int li = 0, ri = 0, i = 0;
        int[] result = new int[nums.length];
        while (li < left.length && ri < right.length) {
            if (left[li] < right[ri]) {
                result[i] = left[li];
                li++;
            } else if (right[ri] < left[li]) {
                result[i] = right[ri];
                ri++;
            } else {
                result[i] = left[li];
                i++;
                result[i] = right[ri];
                li++;
                ri++;
            }
            i++;
        }
        while (li < left.length) {
            result[i] = left[li];
            i++;
            li++;
        }
        while (ri < right.length) {
            result[i] = right[ri];
            i++;
            ri++;
        }
        return Arrays.copyOf(result, i);
    }
}

File: LeetCode (en comments)/912. Sort an Array.cpp
class Solution {
public:
    vector<int> sortArray(vector<int>& nums) {
        if (nums.size() < 2) {
            return nums;
        } else if (nums.size() == 2) {
            if (nums[0] > nums[1]) {
                return {nums[1], nums[0]};
            } else {
                return nums;
            }
        }
        vector<int> left(nums.begin(), nums.begin() + nums.size() / 2);
        vector<int> right(nums.begin() + nums.size() / 2, nums.end());
        left = sortArray(left);
        right = sortArray(right);
        int li = 0, ri = 0;
        vector<int> result;
        while (li < left.size() && ri < right.size()) {
            if (left[li] < right[ri]) {
                result.push_back(left[li]);
                li++;
            } else if (right[ri] < left[li]) {
                result.push_back(right[ri]);
                ri++;
            } else {
                result.push_back(left[li]);
                result.push_back(right[ri]);
                li++;
                ri++;
            }
        }
        while (li < left.size()) {
            result.push_back(left[li]);
            li++;
        }
        while (ri < right.size()) {
            result.push_back(right[ri]);
            ri++;
        }
        return result;
    }
};